Mathlib/Computability/PDA.lean

```diff
@@ -0,0 +1,69 @@
+/-
+Copyright (c) 2025 Stefan Hetzl. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tobias Leichtfried, Stefan Hetzl
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Pushdown Automata
+
+This file contains the definition of a Pushdown Automaton (PDA). Pushdown automata recognize
+exactly the context-free languages.
+-/
+
+/-- A PDA is a set of states (`Q`), a tape alphabet (`T`), a stack alphabet (`S`), an initial state
+  (`initial_state`), a start symbol (`start_symbol`), a set of final states (`final_states`), a
+  transition function (`transition_fun`), and a proof that, for any input, the transition function
+  returns a finite set as output. -/
+structure PDA (Q T S : Type) [Fintype Q] [Fintype T] [Fintype S] where
+  /-- Initial state. -/
+  initial_state : Q
+  /-- Start symbol on the stack. -/
+  start_symbol : S
+  /-- Set of final states. -/
+  final_states : Set Q
+  /-- Transition function reading a letter from `T` if present and an epsilon otherwise. -/
+  transition_fun : Q → (Option T) → S → Set (Q × List S)
+  finite (q : Q) (x : (Option T)) (Z : S) : (transition_fun q x Z).Finite
```

Comment:
Have you considering bundling those two fields into
```suggestion
  /-- Transition function reading a letter from `T` if present and an epsilon otherwise. -/
  transition_fun : Q → (Option T) → S → Finset (Q × List S)
```

Reply:
My student @TobiasLeichtfried has done this formalisation. IIRC he thought about using Finset but it turned out that using .Finite was considerably simpler. I am not sure what would be preferred in this context.

Reply:
> he thought about using Finset but it turned out that using .Finite was considerably simpler. I am not sure what would be preferred in

It would be good to know exactly what was painful with finsets. Finsets are possibly not easy to work with originally, but their use pays off.